### ISSUE TITLE
Consider adding gearboxJoint elements to the hand

### ISSUE THREAD

**Issue by hamalMarino** [CentroEPiaggio/pisa-iit-soft-hand#9]
[Gearbox joints](https://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1physics_1_1GearboxJoint.html) are already present in Gazebo, but cannot be created directly from URDF.
Check whether it is possible to automatically include them in the SDF.

Also keep an eye on [this thread](https://bitbucket.org/osrf/gazebo/issues/1058/update-gearbox-joint-definition) for possible solutions.

@arocchi @carlosjoserg

**Comment by arocchi** [comment]
Added in https://github.com/arocchi/pisa-iit-soft-hand/commit/2421d1d5f8a4ef1e85f074206a95015915d10ae5, need a proper hand loader which select joints based on name rather than on number, since the gearboxes add new joints to the model

**Comment by arocchi** [comment]
After some testing, it appears like the joint handling is actually correct. @carlosjoserg @hamalMarino could you take a look so that we can brainstorm together about the weird effects showing up in simulation?

**Comment by carlosjoserg** [comment]
Should we use https://github.com/arocchi/pisa-iit-soft-hand/commit/2421d1d5f8a4ef1e85f074206a95015915d10ae5 ?

**Comment by arocchi** [comment]
Yes @carlosjoserg . Upon further testing, the problems appear also WITHOUT gearbox and mimic, when the joint damping is too high (e.g. 1.0). It appears like, on instantiation, joints have very high velocities.

**Comment by arocchi** [comment]
One element which I am investigating at the moment is the joint axis. Let's take for example the thumb outer joint. In the SDF, the joint axis for both the normal and the `mimic` joints are:

```
<xyz>0.910366 0.155013 -0.383671</xyz>
```

While in the URDF, it is

```
<joint axis="0 -1 0">
```

When inserting the gearbox joint in the URDF, it gets inserted via a `<gazebo>` tag, meaning that it will get copied over to the SDF without modifications. If the tag `<use_parent_model_frame>` is not specified, it defaults to 1.
So I tried to change the joint axis in the URDF manually for the `gearbox` joint, to the same value I found in the SDF for the `mimic` and the regular joint.
Even with the change, i still didn't get it working. This experimentation has not been successful.
An equivalent alternative to changing the axis of the `gearbox` joint to the one found in the SDF is specifying that

```
<use_parent_model_frame>0</use_parent_model_frame>
```

where the default is 1. At that point, the axis should remain "local", i.e.

```
<xyz>0 -1 0</xyz>
```

I tried also this, and apparently it didn't change results.

**Comment by arocchi** [comment]
Another interesting result is that the gearbox "works" when the gearbox ratio is positive instead of negative, meaning that the phalanx will close automatically (whereas it should stay open) when the synergy is not active. The opposite is not true, that is, a correct gearbox ratio of -1.0 will result in the phalanx getting stuck for every synergy value.

**Comment by arocchi** [comment]
One idea that I also investigated lies in the possibility that applying joint effort does not propagate correctly on gearboxes on 4.1. I made a test https://gist.github.com/arocchi/3f6f770a1f3c33ee178c on  https://bitbucket.org/arocchi/gazebo/branch/gazebo_4.1 that uses joint efforts, and everything appears to work properly. Any news on your side @carlosjoserg @hamalMarino?

**Comment by carlosjoserg** [comment]
The positiveness of the ratio is certainly an issue... is there a diffGearBox in ODE?

**Comment by arocchi** [comment]
ps: note that the positive ratio was "wrong" by design, meaning, it was a test to check the rotation axis was correct...

**Comment by arocchi** [comment]
RE the  `<use_parent_model_frame>0</use_parent_model_frame>` approach: just found out the value get overwritten, or even better, does not get read, so in the final sdf you will end up with two lines for each gearbox axis reading something like this:

```
<use_parent_model_frame>0</use_parent_model_frame>
<use_parent_model_frame>1</use_parent_model_frame>
```

If confirmed, that is a bug that needs to be reported to the gazebo guys

**Comment by arocchi** [comment]
Continuing with tests. Trying to apply the gearbox just on one joint. 
I get instabilities which should be more accurately investigated.
The models under test are in this gist https://gist.github.com/arocchi/52a322c5b39fd99608ed
